Multiply defined symbols when building under cygwin/cygwin64, gcc 6.4 and 7.3 link=shared

I am trying to build iostreams which links against regex dynamically in the Jamfile(s).   Only on cygwin and cygwin64, I get this during CI builds:
```
gcc.link.dll bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\libboost_regex-gcc64-x64-1_68.dll.a
bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\instances.o:instances.cpp:(.text$_ZN5boost16re_detail_10680012perl_matcherIPKcSaINS_9sub_matchIS3_EEENS_12regex_traitsIcNS_14c_regex_traitsIcEEEEE4findEv[_ZN5boost16re_detail_10680012perl_matcherIPKcSaINS_9sub_matchIS3_EEENS_12regex_traitsIcNS_14c_regex_traitsIcEEEEE4findEv]+0x0): multiple definition of `boost::re_detail_106800::perl_matcher<char const*, std::allocator<boost::sub_match<char const*> >, boost::regex_traits<char, boost::c_regex_traits<char> > >::find()'
bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\cregex.o:cregex.cpp:(.text+0x2db0): first defined here
bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\instances.o:instances.cpp:(.text$_ZN5boost16re_detail_10680012perl_matcherIN9__gnu_cxx17__normal_iteratorIPKcSsEESaINS_9sub_matchIS6_EEENS_12regex_traitsIcNS_14c_regex_traitsIcEEEEE4findEv[_ZN5boost16re_detail_10680012perl_matcherIN9__gnu_cxx17__normal_iteratorIPKcSsEESaINS_9sub_matchIS6_EEENS_12regex_traitsIcNS_14c_regex_traitsIcEEEEE4findEv]+0x0): multiple definition of `boost::re_detail_106800::perl_matcher<__gnu_cxx::__normal_iterator<char const*, std::string>, std::allocator<boost::sub_match<__gnu_cxx::__normal_iterator<char const*, std::string> > >, boost::regex_traits<char, boost::c_regex_traits<char> > >::find()'
bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\cregex.o:cregex.cpp:(.text+0x4b90): first defined here
bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\instances.o:instances.cpp:(.text$_ZN5boost16re_detail_10680012perl_matcherIPKcSaINS_9sub_matchIS3_EEENS_12regex_traitsIcNS_14c_regex_traitsIcEEEEE5matchEv[_ZN5boost16re_detail_10680012perl_matcherIPKcSaINS_9sub_matchIS3_EEENS_12regex_traitsIcNS_14c_regex_traitsIcEEEEE5matchEv]+0x0): multiple definition of `boost::re_detail_106800::perl_matcher<char const*, std::allocator<boost::sub_match<char const*> >, boost::regex_traits<char, boost::c_regex_traits<char> > >::match()'
bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\cregex.o:cregex.cpp:(.text+0x3060): first defined here
collect2: error: ld returned 1 exit status
    "g++"   "-Wl,--out-implib,bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\libboost_regex-gcc64-x64-1_68.dll.a" -o "bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\cygboost_regex-gcc64-x64-1_68.dll" -Wl,-h -Wl,cygboost_regex-gcc64-x64-1_68.dll -shared -Wl,--start-group "bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\c_regex_traits.o" "bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\cpp_regex_traits.o" "bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\cregex.o" "bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\fileiter.o" "bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\icu.o" "bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\instances.o" "bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\posix_api.o" "bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\regex.o" "bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\regex_debug.o" "bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\regex_raw_buffer.o" "bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\regex_traits_defaults.o" "bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\static_mutex.o" "bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\w32_regex_traits.o" "bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\wc_regex_traits.o" "bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\wide_posix_api.o" "bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\winstances.o" "bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\usinstances.o"  -Wl,-Bstatic  -Wl,-Bdynamic  -Wl,--end-group -m64 -std=c++03 
...failed gcc.link.dll bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\libboost_regex-gcc64-x64-1_68.dll.a bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\cygboost_regex-gcc64-x64-1_68.dll...
...removing bin.v2\libs\regex\build\gcc-6.4.0\release\cxxstd-03-iso\target-os-cygwin\libboost_regex-gcc64-x64-1_68.dll.a
```
Example log: https://ci.appveyor.com/project/jeking3/iostreams/build/1.0.8-develop/job/tdn544wj0iifm7tw?fullLog=true#L2252

I am trying a quick workaround (just to get iostreams building properly) by adding link=static to the command line and will update this on the result.  I didn't want to lose the issue.

I've pushed a fix for this - most of tests fail on cygwin though as Boost.Thread appears not to be building on cygwin:

```
gcc.compile.c++ ..\..\..\bin.v2\libs\regex\test\regex_regress_threaded.test\gcc-7.3.0cygwin\debug\cxxstd-03-iso\target-o
s-cygwin\threadapi-win32\threading-multi\regress\basic_tests.o
In file included from ..\..\../boost/thread/win32/once.hpp:19:0,
                 from ..\..\../boost/thread/once.hpp:21,
                 from regress\info.hpp:26,
                 from regress\test_not_regex.hpp:22,
                 from regress\test.hpp:31,
                 from regress\basic_tests.cpp:27:
..\..\../boost/thread/win32/thread_primitives.hpp: In function ‘bool boost::detail::win32::interlocked_bit_test_and_set(
long int*, long int)’:
..\..\../boost/thread/win32/thread_primitives.hpp:392:90: error: cannot convert ‘long int*’ to ‘volatile int*’ for argum
ent ‘1’ to ‘int _InterlockedCompareExchange(volatile int*, int, int)’
                     long const current=BOOST_INTERLOCKED_COMPARE_EXCHANGE(x,old|value,old);
                                                                                          ^
..\..\../boost/thread/win32/thread_primitives.hpp: In function ‘bool boost::detail::win32::interlocked_bit_test_and_rese
t(long int*, long int)’:
..\..\../boost/thread/win32/thread_primitives.hpp:409:91: error: cannot convert ‘long int*’ to ‘volatile int*’ for argum
ent ‘1’ to ‘int _InterlockedCompareExchange(volatile int*, int, int)’
                     long const current=BOOST_INTERLOCKED_COMPARE_EXCHANGE(x,old&~value,old);
                                                                                           ^
```

Did you see this also?

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -23,13 +23,15 @@ environment:
     - ARGS: --toolset=msvc-14.0 address-model=32
     - ARGS: --toolset=msvc-12.0 address-model=64
     - ARGS: --toolset=msvc-14.0 address-model=64
-    - ARGS: --toolset=msvc-14.0 address-model=64 cxxflags=-std:c++latest
+    - ARGS: --toolset=msvc-14.0 address-model=64 cxxstd=latest
     - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2017
       ARGS: --toolset=msvc-14.1 address-model=64
     - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2017
       ARGS: --toolset=msvc-14.1 address-model=32
     - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2017
-      ARGS: --toolset=msvc-14.1 address-model=64 cxxflags=-std:c++latest
+      ARGS: --toolset=msvc-14.1 address-model=64 cxxstd=17
+    - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2017
+      ARGS: --toolset=msvc-14.1 address-model=64 cxxstd=latest cxxflags=-permissive-
     - ARGS: --toolset=gcc address-model=64
       PATH: C:\mingw-w64\x86_64-6.3.0-posix-seh-rt_v5-rev1\mingw64\bin;%PATH%
     - ARGS: --toolset=gcc address-model=64 cxxflags=-std=gnu++1z
@@ -38,7 +40,10 @@ environment:
       PATH: C:\mingw-w64\i686-5.3.0-posix-dwarf-rt_v4-rev0\mingw32\bin;%PATH%
     - ARGS: --toolset=gcc address-model=32 linkflags=-Wl,-allow-multiple-definition
       PATH: C:\MinGW\bin;%PATH%
-
+    - ARGS: --toolset=gcc address-model=32 define=_POSIX_C_SOURCE=200112L threadapi=pthread link=static
+      PATH: C:\cygwin\bin;%PATH%
+    - ARGS: --toolset=gcc address-model=64 define=_POSIX_C_SOURCE=200112L define=__USE_ISOC99 threadapi=pthread link=static
+      PATH: C:\cygwin64\bin;%PATH%
 
 install:
   - cd ..

